va/va.go

```diff
@@ -374,7 +384,14 @@ func (va VAImpl) validateTLSALPN01(task *vaTask) *core.ValidationRecord {
 	leafCert := certs[0]
 
 	// Verify SNI - certificate returned must be issued only for the domain we are verifying.
-	if len(leafCert.DNSNames) != 1 || !strings.EqualFold(leafCert.DNSNames[0], task.Identifier) {
+	namematch := false
+	switch task.IdentifierType {
+	case acme.IdentifierDNS:
+		namematch = len(leafCert.DNSNames) == 1 && strings.EqualFold(leafCert.DNSNames[0], task.Identifier)
+	case acme.IdentifierIP:
+		namematch = len(leafCert.IPAddresses) == 1 && leafCert.IPAddresses[0].Equal(net.ParseIP(task.Identifier))
```

Comment:
I'm slightly nervous about this `namematch` implementation for IP addresses. `net.ParseIP` can return `nil` if the IP is invalid. Can you document the assumption that the `task.Identifier` value has already been validated or handle the case where the `parse` returns nil?